meshnet/monitor_event_handler.go

```diff
@@ -0,0 +1,139 @@
+package meshnet
+
+import (
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/NordSecurity/nordvpn-linux/config"
+	"github.com/NordSecurity/nordvpn-linux/core/mesh"
+	"github.com/NordSecurity/nordvpn-linux/internal"
+)
+
+var ErrIncorrectCmdlineContent = errors.New("invalid content of cmdline file of /proc")
+
+// FilesharePortAccessController blocks or allows fileshare port when
+// fileshare process stopped or was restarted accordingly.
+type FilesharePortAccessController struct {
+	cm           config.Manager
+	netw         Networker
+	reg          mesh.Registry
+	filesharePID PID
+}
+
+func NewPortAccessController(cm config.Manager, netw Networker, reg mesh.Registry) FilesharePortAccessController {
+	return FilesharePortAccessController{
+		cm:           cm,
+		netw:         netw,
+		reg:          reg,
+		filesharePID: 0,
+	}
+}
+
+func (eh *FilesharePortAccessController) OnProcessStarted(ev ProcEvent) {
+	if eh.filesharePID != 0 {
+		// fileshare already started and we noted the PID, no need to
+		// process next events anymore until the PID gets reset in [EventHandler.OnProcessStopped]
+		return
+	}
+	if !isFileshareProcess(ev.PID) {
+		return
+	}
+	eh.filesharePID = ev.PID
+	go eh.allowFileshare()
+}
+
+func (eh *FilesharePortAccessController) allowFileshare() error {
+	log.Println(internal.InfoPrefix, "allowing fileshare port")
+	peers, err := eh.listPeers()
+	if err != nil {
+		return err
+	}
+
+	for _, peer := range peers {
+		peerUniqAddr := UniqueAddress{UID: peer.PublicKey, Address: peer.Address}
+		if err := eh.netw.AllowFileshare(peerUniqAddr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (eh *FilesharePortAccessController) listPeers() (mesh.MachinePeers, error) {
+	var cfg config.Config
+	if err := eh.cm.Load(&cfg); err != nil {
+		return nil, fmt.Errorf("reading configuration when listing peers: %w", err)
+	}
+
+	if cfg.MeshDevice == nil {
+		return nil, fmt.Errorf("meshnet is not configured")
+	}
+
+	token := cfg.TokensData[cfg.AutoConnectData.ID].Token
+	peers, err := eh.reg.List(token, cfg.MeshDevice.ID)
+	if err != nil {
+		return nil, fmt.Errorf("listing peers: %w", err)
+	}
+	return peers, nil
+}
+
+func (eh *FilesharePortAccessController) OnProcessStopped(ev ProcEvent) {
+	if eh.filesharePID != ev.PID {
+		return
+	}
+	eh.filesharePID = 0
+	go eh.blockFileshare()
+}
+
+func (eh *FilesharePortAccessController) blockFileshare() error {
+	log.Println(internal.InfoPrefix, "blocking fileshare port")
+	peers, err := eh.listPeers()
+	if err != nil {
+		return err
+	}
+
+	for _, peer := range peers {
+		peerUniqAddr := UniqueAddress{UID: peer.PublicKey, Address: peer.Address}
+		if err := eh.netw.BlockFileshare(peerUniqAddr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func isFileshareProcess(pid PID) bool {
+	// ignore older processes, fileshare is always
+	// younger than the daemon so it has higher PID
+	if pid < PID(os.Getpid()) {
```

Comment:
Is this condition always true? e.g. main daemon crashed/killed then restarted and fileshare daemon was running - will it be restarted as well?

Reply:
When the daemon get's killed, `nordfileshare` also dies. I'll check the crash case to make sure that it isn't graceful shutdown or something similar which causes fileshare to die :+1:

Reply:
When daemon gets killed, the fileshare process also dies, but for the crash case there is no cleanup, so the process is still running. In such case daemon will be younger after restart. I changed the implementation and now, when the event handler gets created, I'm checking first if the fileshare process is already running, if it is - then we save it's PID and this condition won't be executed because if we already have fileshare PID, then the whole `OnProcessStarted` skips the processing.
If, however the PID is not yet saved, then it means that fileshare was not running, so in such case it will always be younger. I added comment describing this [here](https://github.com/NordSecurity/nordvpn-linux/pull/713/files#diff-f91276f6430f8c75462b7aae0000f9b0d50f1d52777c13eb72fc3cd08456211aR57-R62)